app/models/reports/repository.rb

```diff
@@ -0,0 +1,100 @@
+module Reports
+  class Repository
+    def initialize(regions, periods:)
+      @regions = Array(regions)
+      @regions_by_id = @regions.map { |r| [r.id, r] }.to_h
+
+      @periods = if periods.is_a?(Period)
+        Range.new(periods, periods).to_a
+      else
+        periods.to_a
+      end
+    end
+
+    attr_reader :regions, :periods
+    attr_reader :regions_by_id
+    delegate :cache, :logger, to: Rails
+
+    class CacheKey
+      def initialize(key)
+        @key = key
+        @parts = key.split("/")
+      end
+
+      def region_id
+        @parts[1]
+      end
+
+      def slug
+        @parts[2]
+      end
+
+      def period
+        Period.new(type: @parts[3], value: @parts[4])
+      end
+    end
+
+    def lookup_region(id)
+      @regions_by_id.fetch(id)
+    end
+
+    def controlled_patients_info
+      keys = cache_keys(:controlled_patients_info)
+      cached_results = cache.fetch_multi(*keys) { |key|
```

Comment:
This does the Redis MGET -- any keys **not** found in the cache is yielded to the block.